File: apps/api/src/routes/redirect.ts
import { Router } from "express";
import { prisma } from "../lib/prisma.js";
import { redis, REDIS_KEYS, REDIS_TTL } from "../lib/redis.js";
import { notFound } from "../lib/errors.js";
import { incrementClicks } from "../lib/links.js";

export const redirectRouter = Router();

redirectRouter.get("/:code", async (req, res) => {
  const { code } = req.params;

  const cached = await redis.get(REDIS_KEYS.shortCode(code));
  if (cached) {
    incrementClicks(code).catch((err) =>
      console.error("[redirect] Failed to increment clicks:", err)
    );
    return res.redirect(302, cached);
  }

  const link = await prisma.link.findUnique({ where: { code } });
  if (!link) {
    return notFound(res, code);
  }

  await redis.set(REDIS_KEYS.shortCode(code), link.originalUrl, "EX", REDIS_TTL.shortCode);
  await incrementClicks(code);

  return res.redirect(302, link.originalUrl);
});

File: apps/api/routes/redirect.py
from fastapi import APIRouter, Depends, BackgroundTasks
from fastapi.responses import RedirectResponse, JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update

from database import get_db, AsyncSessionLocal
from models import Link
from redis_client import redis, short_code_key, REDIS_TTL_SHORT_CODE

router = APIRouter()


async def _increment_clicks(code: str) -> None:
    async with AsyncSessionLocal() as db:
        await db.execute(update(Link).where(Link.code == code).values(clicks=Link.clicks + 1))
        await db.commit()


@router.get("/{code}")
async def redirect(code: str, background_tasks: BackgroundTasks, db: AsyncSession = Depends(get_db)):
    cached = await redis.get(short_code_key(code))
    if cached:
        background_tasks.add_task(_increment_clicks, code)
        return RedirectResponse(url=cached, status_code=302)

    result = await db.execute(select(Link).where(Link.code == code))
    link = result.scalar_one_or_none()
    if not link:
        return JSONResponse(
            status_code=404,
            content={"error": "NOT_FOUND", "message": f"Short code '{code}' not found"},
        )

    await redis.set(short_code_key(code), link.original_url, ex=REDIS_TTL_SHORT_CODE)
    await db.execute(update(Link).where(Link.code == code).values(clicks=Link.clicks + 1))
    await db.commit()

    return RedirectResponse(url=link.original_url, status_code=302)
